execution_engine/src/core/engine_state/mod.rs

```diff
@@ -641,104 +742,55 @@ where
                                 .write(Key::Account(public_key), StoredValue::Account(new_account))
                         }
                         None => {
+                            // Implies that execution_result is a failure.
                             return Ok(execution_result);
                         }
                     }
                 }
             },
-            Err(error) => {
-                return Ok(ExecutionResult::Failure {
-                    error,
-                    effect: Default::default(),
-                    transfers: Vec::default(),
-                    cost: Gas::default(),
-                });
-            }
+            Err(error) => return Ok(make_charged_execution_failure(error)),
         }
 
-        // Construct a payment code that will put cost of wasmless payment into payment purse
-        let payment_result = {
-            let transfer_args = match runtime_args_builder.clone().build(
-                &account,
-                correlation_id,
-                Rc::clone(&tracking_copy),
-            ) {
+        let transfer_args =
+            match runtime_args_builder.build(&account, correlation_id, Rc::clone(&tracking_copy)) {
                 Ok(transfer_args) => transfer_args,
-                Err(error) => {
-                    return Ok(ExecutionResult::Failure {
-                        error,
-                        effect: Default::default(),
-                        transfers: Vec::default(),
-                        cost: Gas::default(),
-                    });
-                }
+                Err(error) => return Ok(make_charged_execution_failure(error)),
             };
 
+        // Construct a payment code that will put cost of wasmless payment into payment purse
+        let payment_result = {
             // Check source purses minimum balance
-
             let source_uref = transfer_args.source();
 
-            let source_purse_balance_key = match tracking_copy
-                .borrow_mut()
-                .get_purse_balance_key(correlation_id, Key::URef(source_uref))
-            {
-                Ok(purse_balance_args) => purse_balance_args,
-                Err(error) => {
-                    return Ok(ExecutionResult::Failure {
-                        error: Error::Exec(error),
-                        effect: Default::default(),
-                        transfers: Vec::default(),
-                        cost: Gas::default(),
-                    });
-                }
-            };
+            let source_purse_balance = {
+                let source_purse_balance_key = match tracking_copy
```

Comment:
if source purse is main purse, you already have it.

Reply:
Good point! Fixed